### PR TITLE
support Dirac GOS

### PR DESCRIPTION
**Pull request by zezhong-zhang**
### Description of the change
- Support Dirac GOS that includes the relativistic effects as described in the [arxiv](https://arxiv.org/pdf/2405.10151). The dataset is open-source on [Zenodo ](https://zenodo.org/records/12752410) under CC-BY-4.0 license. The format follows the [gosh](https://gitlab.com/gguzzina/gosh)  for the ease of the user, as discussed in https://github.com/hyperspy/hyperspy/pull/3082#issuecomment-1386053704. 
- Currently, I simply duplicate the `exspy/misc/eels/gosh_gos.py` to `exspy/misc/eels/dirac_gos.py` and replace the links. I think it might be better to reduce the redundancy and combine into one. But I hope to hear your preference.
- Some minor changes on the docstring and documentation, to be improved latter upon the feedback of this PR.
- The changelog is not yet done

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] docstring updated (if appropriate),
- [x] update user guide (if appropriate),
- [x] added tests,
- [x] add a changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/exspy/blob/main/upcoming_changes/README.rst)),
- [x] Check formatting of the changelog entry (and eventual user guide changes) in the `docs/readthedocs.org:exspy` build of this PR (link in github checks)
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs

s = hs.load("coreloss_spectrum.msa", signal_type="EELS")
ll = hs.load("lowloss_spectrum.msa", signal_type="EELS")

s.add_elements(("Mn", "O"))
s.set_microscope_parameters(
    beam_energy=300, convergence_angle=24.6, collection_angle=13.6
)
m = s.create_model(low_loss=ll, GOS="dirac") #specify the GOS when create the model 
m.enable_fine_structure()
m.multifit(kind="smart")
m.plot()
```

### CODE REVIEW

**Comment by ericpre** [comment]
Indeed, as you said, the duplication should be removed.

Two approaches can be used:

1. Create two subclass of `GoshGOS`:
    - `GoshGOS` handle the functionalities related to the gosh file, etc.
    - `DFTGOS` (or similar name) is a subclass of `GoshGOS` with the url to download the GOS and the corresponding hash for https://zenodo.org/records/7645765, saved as attribute of the class.
    - `DiracGOS` is the same as `DFTGOS` but for https://zenodo.org/records/12751756
2. The other approach is to define a registry handling the urls and hashes.

@gguzzina, @francisco-dlp, any preferences?

On a side note, saving a model `GoshGOS` with a custom path may not be working currently... because the url and hash are not saved. If going for the registry approach, this may need to be fixed.

**Comment by gguzzina** [comment]
The registry approach seems more sensible to me. If implemented in a sensible way it allows two things:
1) to ship a list of known GOS (also allowed by the subclass method)
2) to use other GOS data without need to edit the sources

The new GOS data could include even revisions of the two datasets above, which might be at different URLs or have different hashes.

**Comment by zezhong-zhang** [comment]
The first approach can be easily done, but have the disadvantage of update the url and hash in the code everytime when update the databases. For the registry approach, do we have an mini example of how it works?

**Comment by ericpre** [comment]
It is already possible to specify a file path for a gosh file: https://hyperspy.org/exspy/user_guide/eels.html#generalised-oscillator-strengths.

In the interest of simplicity and because I don't think that there is a use case for user changing the registry dynamically (the cross section are expected to be stable at some point and change rarely and user can use other gosh file easily), I would suggest to use a registry that is simply a python dictionary hardcoded in exspy, that would need to updated one in a while.

**Comment by zezhong-zhang** [comment]
To check if I understand it correctly, so we use the GoshGOS and add a dictionary of registry, instead of add subclasses. The GoshGOS will then need to have a new parameter (call it "source"?) to specify if use DFT or Dirac, default to DFT. 

Currently, the pooch download the database to a cache directory by default, might be better to put in the database directory within the package.

**Comment by ericpre** [comment]
> To check if I understand it correctly, so we use the GoshGOS and add a dictionary of registry, instead of adding subclasses. The GoshGOS will then need to have a new init parameter (call it "source"?) to specify if use DFT or Dirac, default to DFT.
> 
> The use of the GOS will be
> 
> ```python
> m = s.create_model(low_loss=ll, GOS="Dirac")
> m = s.create_model(low_loss=ll, GOS="DFT") # currently called gosh
> m = s.create_model(low_loss=ll) # this will use dft-gosh by default
> ```

This looks good to me. I agree that `"gosh"` to `"DFT"`, considering that "gosh" refer to the file format and this argument is about specifying the type of GOS. I will defer to you and @gguzzina to define the most suitable name for these two! 😃 

> Currently, the pooch download the database to a cache directory by default, might be better to put in the database directory within the package.

It is good practise to keep the distribution package as small as reasonable and therefore it is better to download and cache the database. This is only be download only once.
Off the top of my head, pypi.org only allows packages ~50 MB and looking at https://zenodo.org/records/12752410, the file is 500 MB. By the way, is it expected that the file is so large? In comparison, the gosh file from https://zenodo.org/records/7645765 is ~42 MB.
If you are thinking of be able to use these GOS easily on air gap system, then it would be better to discuss this in a separate issue.

**Comment by zezhong-zhang** [comment]
What I mean is that the package indeed does not distribute those files and still do the download, but to define a directory that holds the database for downloading, instead of putting in the cache directory by default which user hard to find and likely to purge. 

The Dirac GOS database has all the elements and all the edges (over 2000 entries). In each entry, we have relatively dense sampling (256 in momentum and 128 in energy). Due to spin-orbit coupling, the L2 and L3 for instance are two different entries, instead of single array in the Schrodinger approach. We also used double precision which doubles the size. If preferred, we can give the single precision version but will still be ~200 MB.
> 
> > Currently, the pooch download the database to a cache directory by default, might be better to put in the database directory within the package.
> 
> It is good practise to keep the distribution package as small as reasonable and therefore it is better to download and cache the database. This is only be download only once. Off the top of my head, pypi.org only allows packages ~50 MB and looking at https://zenodo.org/records/12752410, the file is 500 MB. By the way, is it expected that the file is so large? In comparison, the gosh file from https://zenodo.org/records/7645765 is ~42 MB. If you are thinking of be able to use these GOS easily on air gap system, then it would be better to discuss this in a separate issue.

**Comment by ericpre** [comment]
> What I mean is that the package indeed does not distribute those files and still do the download, but to define a directory that holds the database for downloading, instead of putting in the cache directory by default which user hard to find and likely to purge.

Writing to `site-packages` is bad practise for various reasons (user doesn't necessarily have the necessarily privilege, files doesn't get removed when uninstalling, suboptimal for python environment, etc.).
It is possible to use a different [cache folder](https://www.fatiando.org/pooch/latest/api/generated/pooch.retrieve.html#pooch.retrieve) that would differ from the default cache location, but would is the use case? Users are not expected to read the files themselves?

> The Dirac GOS database has all the elements and all the edges (over 2000 entries). In each entry, we have relatively dense sampling (256 in momentum and 128 in energy). Due to spin-orbit coupling, the L2 and L3 for instance are two different entries, instead of single array in the Schrodinger approach. We also used double precision which doubles the size. If preferred, we can give the single precision version but will still be ~200 MB.

Could the file size be reduced by using [compression](https://docs.h5py.org/en/stable/high/dataset.html#lossless-compression-filters)?

**Comment by zezhong-zhang** [comment]
1. Alright, I will not insist on the database path for downloading. 
2. I uploaded a [compact database](https://zenodo.org/records/12800856) with single precision, down-sampling and compression to get the dataset down to ~66 MB, should be relatively easy for user to download. The link in the code is updated accordingly. 
3. The methods can be called in following way:
```python
from exspy.misc.eels.gosh_gos import GoshGOS
GoshGOS('Ti_L2', source='dirac')

from exspy.models.eelsmodel import EELSModel
EELSModel(signal_1d, GOS='dirac')

from exspy.signals.eels import EELSSpectrum

s = EELSSpectrum(data)
s.create_model(GOS='dirac')
```
4. There are breaking changes in the naming: the default is no longer `GOS=gosh` but  `GOS=dft` in the above classes. As 'dft' and 'dirac' will use both the `GoshGOS`, the `_name` will be `dft_gosh` and `dirac_gosh`. @gguzzina, could you have a look of the changes to check if the names are appropriate?

File: upcoming_changes/72.new.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Support for [Dirac GOS](https://zenodo.org/records/12800856) in the gosh format for EELS quantification, which includes the relativistic effects (e.g. spin-orbit coupling) based on Dirac equation.
```
Comment (ericpre) [inline upcoming_changes/72.new.rst:1]: This is a `rst` file and rst formatting needs to be used: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages
```suggestion
Support for `Dirac GOS <https://zenodo.org/records/12800856>`_ in the gosh format for EELS quantification, which includes the relativistic effects (e.g. spin-orbit coupling) based on Dirac equation. 
```

File: exspy/components/eels_cl_edge.py [hunk 2]
```diff
@@ -146,7 +147,7 @@ class EELSCLEdge(Component):
     _fine_structure_coeff_free = True
     _fine_structure_spline_active = True
 
-    def __init__(self, element_subshell, GOS="gosh", gos_file_path=None):
+    def __init__(self, element_subshell, GOS="dft", gos_file_path=None):
```
Comment (ericpre) [inline exspy/components/eels_cl_edge.py:150]: A side note as this is outside of the scope of this PR: I find it confusing to have both `GOS` and `gos_file_path`. I am wondering if these should be merged together to make `GOS` accept one of the supported string (`"dft", "dirac", etc.) or a path (which could a str or `pathlib.Path`) in which case, it will try to read the file assuming this is gosh file?

File: exspy/components/eels_cl_edge.py [hunk 3]
```diff
@@ -167,15 +168,21 @@ def __init__(self, element_subshell, GOS="gosh", gos_file_path=None):
         self.fine_structure_coeff.ext_force_positive = False
         self.GOS = None
 
-        if GOS == "gosh":
-            self.GOS = GoshGOS(element_subshell, gos_file_path=gos_file_path)
+        if GOS == "dft":
```
Comment (ericpre) [inline exspy/components/eels_cl_edge.py:171]:
```suggestion
        if GOS == "gosh":
              warnings.warn(
                "The value 'gosh' of the `GOS` parameter has been renamed to 'dft' in "
                "eXSpy 0.3.0, use `GOS='dft'` instead. "
                "Using `GOS='gosh'` will stop working in eXSpy 1.0.",
                VisibleDeprecationWarning,
            )
            GOS = "dft"
        if GOS == "dft":
```

Corresponding import needs to be added at the beginning of the file:
```python
from hyperspy.exceptions import VisibleDeprecationWarning
```

**Review by ericpre — COMMENTED** [review]
This looks in very good shape.

File: exspy/components/eels_cl_edge.py (hunk 2 above)
Comment (zezhong-zhang) [reply exspy/components/eels_cl_edge.py:150]: Agree, this makes sense to me.

**Comment by zezhong-zhang** [comment]
Not sure why windows test failed due to connectivity, but linux passed ... @ericpre do you have any suggestions?

File: exspy/tests/models/test_eelsmodel.py [hunk 4]
```diff
@@ -73,19 +73,44 @@ def test_gos_hydrogenic(self):
 
     def test_gos_gosh(self):
         m = self.s.create_model(auto_add_edges=True, GOS="gosh")
```
Comment (ericpre) [inline exspy/tests/models/test_eelsmodel.py:75]: To test that it does warn as expected and capture the warning to avoid adding more warning when running the test suite.
```suggestion
        with pytest.warns(VisibleDeprecationWarning):
            m = self.s.create_model(auto_add_edges=True, GOS="gosh")
```

File: exspy/misc/eels/gosh_gos.py [hunk 5]
```diff
@@ -79,20 +87,25 @@ class GoshGOS(TabulatedGOS):
...
         gos_file_path : str
             The path of the gosh file to use.
+        source : str
+            The source of the GOS data. Options are 'dft' or 'dirac'.
         """
 
         if gos_file_path is None:
+            source = source.lower()
+            assert source in GOSH_SOURCES.keys(), f"Invalid source: {source}"
+            self._name = f"{source}_gosh"
```
Comment (ericpre) [inline exspy/misc/eels/gosh_gos.py:105]: Does it need the "gosh" suffix?
```suggestion
            self._name = source
```

**Review by ericpre — APPROVED** [review]
This looks very good to me - 2 minor comments.

@gguzzina, does it sound good to you to rename "gosh" to "dft" to make the distinction because the two GOS as mentioned [above](https://github.com/hyperspy/exspy/pull/72#issuecomment-2245510694)?

**Comment by gguzzina** [comment]
@ericpre that's fine by me.

In my personal use I've started to use author names.
The traditional Hartree-Slater would be Rez, our DFT would be Segger, and the Dirac would be Zhang.

But until there's an invasion of competing DFT and Dirac datasets I don't think we have a problem.

**Comment by zezhong-zhang** [comment]
@gguzzina Indeed the author names would be more specific, but the users probably would not have a instinct guess (can add in the doc string which corresponds the author name with the method). Let's keep this option open when the next new GOS added to the list.
